File: src/main/java/com/gregtechceu/gtceu/api/sync_system/SyncActionContext.java
package com.gregtechceu.gtceu.api.sync_system;

import com.gregtechceu.gtceu.api.cover.CoverBehavior;
import com.gregtechceu.gtceu.api.sync_system.managed.ManagedSyncBlockEntity;

import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.core.component.DataComponentMap;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.item.ItemStack;

import org.jetbrains.annotations.Nullable;

/**
 * Server-side context passed to a registered sync action handler.
 *
 * @param player the server player that sent the action request.
 * @param holder the resolved machine, cover, or item holder for the request.
 * @param action the decoded action data.
 * @param pos    the block position for block-backed holders, or {@code null} for item actions.
 * @param side   the cover side for cover holders, or {@code null} for other holders.
 * @param hand   the interaction hand for held item holders, or {@code null} for block-backed holders.
 */
public record SyncActionContext(ServerPlayer player, Object holder, SyncActionData action, @Nullable BlockPos pos,
                                @Nullable Direction side, @Nullable InteractionHand hand) {

    /**
     * Creates context for a managed block entity action.
     */
    public static SyncActionContext machine(ServerPlayer player, ManagedSyncBlockEntity holder, SyncActionData action,
                                            BlockPos pos) {
        return new SyncActionContext(player, holder, action, pos, null, null);
    }

    /**
     * Creates context for an attached cover action.
     */
    public static SyncActionContext cover(ServerPlayer player, CoverBehavior holder, SyncActionData action, BlockPos pos,
                                          Direction side) {
        return new SyncActionContext(player, holder, action, pos, side, null);
    }

    /**
     * Creates context for an action targeting the item currently held by the player.
     */
    public static SyncActionContext item(ServerPlayer player, ItemStack holder, SyncActionData action,
                                         InteractionHand hand) {
        return new SyncActionContext(player, holder, action, null, null, hand);
    }

    /**
     * Returns the action id without forcing handlers to dereference the action record.
     */
    public ResourceLocation actionId() {
        return action.actionId();
    }

    /**
     * Returns the client sequence number without forcing handlers to dereference the action record.
     */
    public int sequence() {
        return action.sequence();
    }

    /**
     * Returns the component payload that action handlers must validate before use.
     */
    public DataComponentMap payload() {
        return action.payload();
    }
}

File: src/main/java/com/gregtechceu/gtceu/api/sync_system/SyncActionContext.kt
package com.gregtechceu.gtceu.api.sync_system

import com.gregtechceu.gtceu.api.cover.CoverBehavior
import com.gregtechceu.gtceu.api.sync_system.managed.ManagedSyncBlockEntity

import net.minecraft.core.component.DataComponentMap
import net.minecraft.resources.ResourceLocation
import net.minecraft.server.level.ServerPlayer

import net.minecraft.core.BlockPos as Pos
import net.minecraft.core.Direction as Side
import net.minecraft.world.InteractionHand as Hand
import net.minecraft.world.item.ItemStack as Stack

/**
 * Server-side context passed to a registered sync action handler.
 *
 * @property player the server player that sent the action request.
 * @property holder the resolved machine, cover, or item holder for the request.
 * @property action the decoded action data.
 * @property pos the block position for block-backed holders, or `null` for item actions.
 * @property side the cover side for cover holders, or `null` for other holders.
 * @property hand the interaction hand for held item holders, or `null` for block-backed holders.
 * @property openedStack the held item stack snapshot captured when the UI was opened, or `null` for block-backed
 * holders.
 */
@JvmRecord
data class SyncActionContext(val player: ServerPlayer, val holder: Any, val action: SyncActionData, val pos: Pos?, val side: Side?, val hand: Hand?, val openedStack: Stack?) {
	/**
	 * Returns the action id without forcing handlers to dereference the action record.
	 */
	fun actionId(): ResourceLocation = action.actionId

	/**
	 * Returns the client sequence number without forcing handlers to dereference the action record.
	 */
	fun sequence(): Int = action.sequence

	/**
	 * Returns the component payload that action handlers must validate before use.
	 */
	fun payload(): DataComponentMap = action.payload

	override fun toString(): String = "SyncActionContext[player=$player, holder=$holder, action=$action, pos=$pos, side=$side, hand=$hand, openedStack=$openedStack]"

	companion object {
		/**
		 * Creates context for a managed block entity action.
		 */
		@JvmStatic
		fun machine(player: ServerPlayer, holder: ManagedSyncBlockEntity, action: SyncActionData, pos: Pos): SyncActionContext = SyncActionContext(player, holder, action, pos, null, null, null)

		/**
		 * Creates context for an attached cover action.
		 */
		@JvmStatic
		fun cover(player: ServerPlayer, holder: CoverBehavior, action: SyncActionData, pos: Pos, side: Side): SyncActionContext = SyncActionContext(player, holder, action, pos, side, null, null)

		/** Creates context for a held item action. */
		@JvmStatic
		fun item(player: ServerPlayer, holder: Stack, openedStack: Stack, action: SyncActionData, hand: Hand) = SyncActionContext(player, holder, action, null, null, hand, openedStack)
	}
}
